kv: replace SystemSpanConfig with SystemDescriptorTableSpan / SystemZonesTableSpan

RichardJCai opened the pull request:
SystemSpanConfig does not have to contain all tables with id <= 10.
It has been replaced with SystemDescriptorTableSpan and SystemZonesTableSpan.
These two tables are the only system tables that are unsplittable.

We cannot get rid of the concept entirely due to SystemSpanConfig components
being relied on for replication reports and opt catalog.

Release note: None

fixes https://github.com/cockroachdb/cockroach/issues/82178

RichardJCai committed:
kv: replace SystemSpanConfig with SystemDescriptorTableSpan / SystemZonesTableSpan

SystemSpanConfig does not have to contain all tables with id <= 10.
It has been replaced with SystemDescriptorTableSpan and SystemZonesTableSpan.
These two tables are the only system tables that are unsplittable.

We cannot get rid of the concept entirely due to SystemSpanConfig components
being relied on for replication reports and opt catalog.

Release note: None

RichardJCai commented:
Is this the right idea? @ajwerner @irfansharif

ajwerner commented on pkg/server/systemconfigwatcher/systemconfigwatchertest/test_system_config_watcher.go:
```diff
@@ -118,18 +119,37 @@ func runTest(
-		false, // forUpdate
-	))
+	ba.Add(
+		roachpb.NewScan(
+			append(codec.TenantPrefix(), keys.SystemDescriptorTableSpan.Key...),
+			append(codec.TenantPrefix(), keys.SystemDescriptorTableSpan.EndKey...),
+			false, // forUpdate
+		),
+	)
 	br, pErr := kvDB.NonTransactionalSender().Send(ctx, ba)
 	require.NoError(t, pErr.GoError())
-	return br.Responses[0].GetScan().Rows
+
+	rows1 := br.Responses[0].GetScan().Rows
+
+	ba.Reset()
+
+	ba.Add(
+		roachpb.NewScan(
+			append(codec.TenantPrefix(), keys.SystemZonesTableSpan.Key...),
+			append(codec.TenantPrefix(), keys.SystemZonesTableSpan.EndKey...),
+			false, // forUpdate
+		),
+	)
+	br, pErr = kvDB.NonTransactionalSender().Send(ctx, ba)
+	require.NoError(t, pErr.GoError())
+
+	rows2 := br.Responses[0].GetScan().Rows
+
+	return append(rows1, rows2...)
```
make a helper

irfansharif replied:
+1, a helper to construct this `row` thing for each span would help. Would also let you avoid resetting a batch request for re-use.

ajwerner commented on pkg/spanconfig/spanconfigstore/span_store.go:
```diff
@@ -115,11 +115,19 @@ func (s *spanConfigStore) computeSplitKey(start, end roachpb.RKey) (roachpb.RKey
-	if keys.SystemConfigSpan.Contains(sp) {
+	// TODO(richardjcai): Can we remove this?
+	if keys.SystemDescriptorTableSpan.Contains(sp) {
 		return nil, nil
 	}
-	if keys.SystemConfigSpan.ContainsKey(sp.Key) {
-		return roachpb.RKey(keys.SystemConfigSpan.EndKey), nil
+	if keys.SystemDescriptorTableSpan.ContainsKey(sp.Key) {
+		return roachpb.RKey(keys.SystemDescriptorTableSpan.EndKey), nil
+	}
+
+	if keys.SystemZonesTableSpan.Contains(sp) {
+		return nil, nil
+	}
+	if keys.SystemZonesTableSpan.ContainsKey(sp.Key) {
+		return roachpb.RKey(keys.SystemZonesTableSpan.EndKey), nil
```
I think this can go

irfansharif replied:
Yup! All of this can go.

ajwerner commented on pkg/sql/opt_catalog.go:
```diff
@@ -1510,8 +1510,10 @@ func (oi *optIndex) Span() roachpb.Span {
 	// span.
-	if desc.GetID() <= keys.MaxSystemConfigDescID {
-		return keys.SystemConfigSpan
+	if desc.GetID() == keys.DescriptorTableID {
+		return keys.SystemDescriptorTableSpan
+	} else if desc.GetID() == keys.ZonesTableID {
+		return keys.SystemZonesTableSpan
```
I think this can go

ajwerner commented on pkg/kv/kvserver/replica_gossip.go:
```diff
@@ -126,13 +126,19 @@ func (r *Replica) MaybeGossipNodeLivenessRaftMuLocked(
-func (r *Replica) loadSystemConfig(ctx context.Context) (*config.SystemConfigEntries, error) {
+// loadSystemDescriptorAndZonesTableSpans scans the system.descriptor
+// and system.zones spans and the fixed spans for the two tables.
+// TODO(richardjcai): Do we still need this? It is only used in tests.
```
I think this should go

irfansharif replied:
+1.

ajwerner commented on pkg/server/systemconfigwatcher/cache.go:
```diff
@@ -88,16 +88,20 @@ func NewWithAdditionalProvider(
-	span := roachpb.Span{
-		Key:    append(codec.TenantPrefix(), keys.SystemConfigSplitKey...),
-		EndKey: append(codec.TenantPrefix(), keys.SystemConfigTableDataMax...),
+	spans := []roachpb.Span{
```
👍

ajwerner reviewed (commented):
It's the right idea

irfansharif commented on pkg/spanconfig/spanconfigstore/span_store.go:
```diff
@@ -115,11 +115,19 @@ func (s *spanConfigStore) computeSplitKey(start, end roachpb.RKey) (roachpb.RKey
 	// TODO(irfansharif): Once we've fully phased out the system config span, we
 	// can get rid of this special handling.
-	if keys.SystemConfigSpan.Contains(sp) {
+	// TODO(richardjcai): Can we remove this?
```
Yup!

irfansharif commented on pkg/sql/logictest/testdata/logic_test/create_index_on_system_table:
```diff
@@ -0,0 +1,9 @@
+INSERT INTO system.users VALUES ('node', '', false);
+GRANT node TO root;
+
+# Any system table other than system.descriptor and system.zones
```
Why can't `system.descriptor` or `system.zones` be split across multiple ranges?

irfansharif commented on pkg/kv/kvserver/store.go:
```diff
@@ -2100,9 +2100,15 @@ func (s *Store) startGossip() {
 			interval:    s.cfg.SentinelGossipTTL() / 2,
 		},
 		{
-			key:         keys.SystemConfigSpan.Key,
```
Do we need this anymore? We don't need to be gossiping writes to these two tables, right? IIUC, it was only there for compatibility with 21.2 versions, which this code (to-be-released in v22.2) doesn't care about.

irfansharif commented on pkg/kv/kvserver/replica_test.go:
```diff
@@ -6994,7 +6994,7 @@ func TestReplicaLoadSystemConfigSpanIntent(t *testing.T) {
 			return nil
 		}
-		return errors.New("recent write not found in gossiped SystemConfig")
+		return errors.New("recent write not found in gossiped system.descriptor and system.zones spans")
```
See below, I don't think we need to gossip these writes anymore so this test could be trimmed down.

irfansharif commented on pkg/keys/spans.go:
```diff
@@ -43,13 +43,17 @@ var (
-	NoSplitSpans = []roachpb.Span{Meta1Span, Meta2MaxSpan, NodeLivenessSpan, SystemConfigSpan}
+	// SystemDescriptorTableSpan, SystemZonesTableSpan cannot be split due to
+	// still being depended on in the "SystemConfigSpan" which is used for
+	// replication reports and opt catalog.
```
See comment below, I think we should be able to split within these tables. Update this + commentary if so.

irfansharif commented on pkg/keys/printer_test.go:
```diff
@@ -116,7 +116,7 @@ func TestPrettyPrint(t *testing.T) {
 
 		// table
-		{keys.SystemConfigSpan.Key, "/Table/SystemConfigSpan/Start", revertSupportUnknown},
+		{keys.TableDataMin, "/Table/SystemConfigSpan/Start", revertSupportUnknown},
```
Do we need this pretty printing now that this span is no longer distinguished?

irfansharif reviewed (commented):
I need to take a closer look at the spanconfigcompared testdata files. Honestly, that package existed to compare the span configs infra against the old thing, the old thing we've now gotten rid of. I'd be happy nuking that entire test package. Wanna do the honours?

ajwerner commented:
> Honestly, that package existed to compare the span configs infra against the old thing, the old thing we've now gotten rid of. I'd be happy nuking that entire test package. Wanna do the honours?

I'm not convinced we're ready for that. Recall that the optimizer and the replication reports still derive their expectations of what the span configs will be from the pre-existing logic.

RichardJCai commented:
Also anyone have an idea on why `TestStoreRangeUpReplicate` and `TestRangeCount` pass locally for me but not in CI?

RichardJCai committed:
kv: fix tests that count ranges after SystemConfigSpan change

Release note: None

knz reviewed (commented):
I'm still seeing the SystemConfigProvider being used. Is that also meant to go away after this PR?
For example, is the thrust of this change also going to cause kvserver/reports/reporter.go to use the new descriptor/zones?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82281#-:-N4bAQ9rBqZA5swYXTxN:b-r2knfe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @RichardJCai)



<!-- Sent from Reviewable.io -->

ajwerner commented:
>For example, is the thrust of this change also going to cause kvserver/reports/reporter.go to use the new descriptor/zones?

No, that's out of scope for this change. This change is to make it such that the system config span tables can be split and such that the trigger concepts are removed. There are still several components which interact with the `config.SystemConfig` object, but get it through a rangefeed. This is safe for secondary tenants to use for the existing use cases they are needed for, so there's nothing pressing us to change it; the span config infrastructure derives its state from the data in descriptor and zones, so nothing about their logic needs to change. 

The two primary users of the SystemConfigProvider are the optimizer and the replication reports.

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82281)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @RichardJCai)

___
*[`pkg/keys/spans.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82281#-N3WDkUYQ9T2-lsIgWZ6-r1-57:-N48k5OyBgK6cKKhespM:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/921b11c3bc2dbcf665a3d9c89069127cb7f9ab88/pkg/keys/spans.go#L57)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

See comment below, I think we should be able to split within these tables. Update this + commentary if so.
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/replica_gossip.go` line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82281#-N3Usi8HWdVIZDP0lMzf-r1-131:-N48lD3z9kgrpX7DRjd1:b-vrq0j7) ([raw file](https://github.com/cockroachdb/cockroach/blob/921b11c3bc2dbcf665a3d9c89069127cb7f9ab88/pkg/kv/kvserver/replica_gossip.go#L131)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

+1.
</blockquote></details>

Deleted and deleted `TestReplicaLoadSystemConfigSpanIntent`.
___
*[`pkg/spanconfig/spanconfigstore/span_store.go` line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82281#-N3WDkUYQ9T2-lsIgWZG-r1-118:-N48lPScDSPBZXuu9Ew8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/921b11c3bc2dbcf665a3d9c89069127cb7f9ab88/pkg/spanconfig/spanconfigstore/span_store.go#L118)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Yup!
</blockquote></details>

Done.
___
*[`pkg/spanconfig/spanconfigstore/span_store.go` line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82281#-N3WDkUYQ9T2-lsIgWZG-r1-130:-N48lPpw7CpofGatKAUT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/921b11c3bc2dbcf665a3d9c89069127cb7f9ab88/pkg/spanconfig/spanconfigstore/span_store.go#L130)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Yup! All of this can go.
</blockquote></details>

Done.
___
*[`pkg/sql/opt_catalog.go` line 1516 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82281#-N3WDkUYQ9T2-lsIgWZJ-r1-1516:-N48lWCt01zdodxBPq4h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/921b11c3bc2dbcf665a3d9c89069127cb7f9ab88/pkg/sql/opt_catalog.go#L1516)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think this can go 
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/create_index_on_system_table` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82281#-N3fZrYxdc8hNrn4B8Zw-r1-5:-N48lYf33KDk3asn5NxA:buo4ijo) ([raw file](https://github.com/cockroachdb/cockroach/blob/921b11c3bc2dbcf665a3d9c89069127cb7f9ab88/pkg/sql/logictest/testdata/logic_test/create_index_on_system_table#L5)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Why can't `system.descriptor` or `system.zones` be split across multiple ranges?
</blockquote></details>

Copied the remaining "system config span" tables into the unsplittable range but I don't believe that is the case anymore.
___
*[`pkg/keys/printer_test.go` line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82281#-N440Bo385WZh6WnsvvL-r1-119:-N48nqfM-lxD8EEhWgzm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/921b11c3bc2dbcf665a3d9c89069127cb7f9ab88/pkg/keys/printer_test.go#L119)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Do we need this pretty printing now that this span is no longer distinguished?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica_test.go` line 6997 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82281#-N3Usi8HWdVIZDP0lMzh-r1-6997:-N48ocT54JqdH8q7GXPi:bhofej8) ([raw file](https://github.com/cockroachdb/cockroach/blob/921b11c3bc2dbcf665a3d9c89069127cb7f9ab88/pkg/kv/kvserver/replica_test.go#L6997)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

See below, I don't think we need to gossip these writes anymore so this test could be trimmed down.
</blockquote></details>

Deleted


<!-- Sent from Reviewable.io -->

RichardJCai commented:
> spanconfigcompared 

Also I deleted this package, I updated some of the range tests, the one thats kind of suspect to me is `TestStoreRangeUpReplicate` in `client_raft_test.go`

ajwerner reviewed (approved):
LGTM

Thank you for doing this. It makes me very happy.

RichardJCai commented on pkg/sql/logictest/testdata/logic_test/create_index_on_system_table:
```diff
@@ -0,0 +1,9 @@
+statement ok
+INSERT INTO system.users VALUES ('node', '', false);
+GRANT node TO root;
+
+# System tables can have their ranges split. Previously system tables
+# with ID <= 10 could not be split.
+
+statement ok
+CREATE INDEX ON system.users ("username");
```
I think this test actually doesn't provide much value, I'll just remove it.

irfansharif commented on pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go:
```diff
@@ -366,70 +366,27 @@ func (s *SQLTranslator) generateSpanConfigurationsForTable(
 
 		return records, nil
-
-		// TODO(irfansharif): There's an attack vector here that we haven't
-		// addressed satisfactorily. By splitting only on start keys of span
-		// configs, a malicious tenant could augment their reconciliation
-		// process to install configs starting much later in their addressable
-		// keyspace. This could induce KV to consider a range boundary that
-		// starts at the previous tenant's keyspace (albeit the tail end of it)
-		// and ending within the malicious tenant's one -- that's no good. We
-		// could do two things:
-		// (i)  Have each tenant install a span config that demarcates the end of
-		//      its keyspace (in our example the previous tenant would defensively
-		//      prevent leaking its user data through this hard boundary);
-		// (ii) Have KV enforce these hard boundaries in with keyspans that
-		//      straddle tenant keyspaces.
-		//
-		// Doing (ii) feels saner, and we already do something similar when
-		// seeding `system.span_configurations` for newly created tenants. We
-		// could have secondary tenants still govern span configurations over
-		// their keyspace, but we'd always split on the tenant boundary. For
-		// malicious tenants, the config we'd apply over that range would be the
-		// fallback one KV already uses for missing spans. For non-malicious
-		// tenants, it could either be the config for (i) `system.descriptor` as
-		// done above, or (ii) whatever the tenant's RANGE DEFAULT is.
-		//
-		// See #73749.
```
Yup, still applicable. Do revert (though I don't know if we'll ever do such things, like most of my TODOs.)

irfansharif commented on pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go:
```diff
@@ -366,70 +366,27 @@ func (s *SQLTranslator) generateSpanConfigurationsForTable(
-			}
-			records = append(records, record)
-		} else {
-			// The same as above, except we have named ranges preceding
```
Perhaps reword some of this commentary to explain why for the system tenant we're choosing to split at SystemConfigSplitKey. I'll actually suggest using TableDataMin here instead (removing another occurrence of "SystemConfig" from SystemConfigSplitKey).

irfansharif reviewed (approved):
Really great stuff!

irfansharif commented on pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go:
```diff
@@ -366,43 +366,34 @@ func (s *SQLTranslator) generateSpanConfigurationsForTable(
+		// splitting on /Table/3 instead of /Table/0 which is benign since there's
+		// no data under /Table/{0-2}.
+		// Still, doing it this way reduces the
+		// differences between the gossip-backed subsystem and this one --
```
nit: reword to remove the reference to "gossip-backed subsystem", or comparisons to subsystems. Just say that we have named ranges (liveness, meta) before the first table in the system tenant keyspace, so we pick the first table ID. That doesn't apply to the tenant keyspace so we can simply split at the tenant prefix itself.

RichardJCai commented:
Alright I think I addressed everything, thanks for the reviews!

bors r+

RichardJCai commented:
bors r+
